rct_optimizations/test/pnp_utest.cpp

```diff
@@ -0,0 +1,81 @@
+#include <rct_optimizations/experimental/pnp.h>
+#include <gtest/gtest.h>
+#include <rct_optimizations_tests/utilities.h>
+#include <rct_optimizations_tests/observation_creator.h>
+
+using namespace rct_optimizations;
+
+TEST(PnPTest, 2DValidation)
+{
+  //setup params
+  PnPProblem setup;
+
+  //Load camera intrinsics
+  //camera is a kinect
+  auto camera = test::makeKinectCamera();
+  setup.intr = camera.intr;
+
+  //testing with only one observation of a 5x5 target
+  //make target
+  test::Target target(5, 5, 0.025);
+
+  Eigen::Isometry3d target_point = Eigen::Isometry3d::Identity();
+  Eigen::Isometry3d camera_point = Eigen::Isometry3d::Identity();
+
+  camera_point.translate(Eigen::Vector3d(0.0,0.0,1.0));
+  camera_point.rotate(Eigen::AngleAxisd(M_PI,Eigen::Vector3d::UnitX()));
+
+  EXPECT_NO_THROW(
+    setup.correspondences = getCorrespondences(camera_point,
+                                               target_point,
+                                               camera,
+                                               target,
+                                               true);
+  );
+
+  setup.camera_to_target_guess = camera_point;
+
+  PnPResult res = rct_optimizations::optimize(setup);
+  EXPECT_TRUE(res.converged);
+  EXPECT_TRUE(res.initial_cost_per_obs < 1.0e-13);
+  EXPECT_TRUE(res.final_cost_per_obs < 1.0e-13);
+  EXPECT_TRUE(res.camera_to_target.isApprox(setup.camera_to_target_guess));
+}
+
+TEST(PnPTest, 3DValidation)
+{
+  //setup params
+  PnPProblem3D setup;
+
+  //testing with only one observation of a 3x3 target
+
+  //make target
+  test::Target target(5, 5, 0.025);
+
+  Eigen::Isometry3d target_point = Eigen::Isometry3d::Identity();
+  Eigen::Isometry3d camera_point = target_point;
+
+  camera_point.translate(Eigen::Vector3d(0.0,0.0,1.0));
+  camera_point.rotate(Eigen::AngleAxisd(M_PI,Eigen::Vector3d::UnitX()));
+
+  setup.correspondences = getCorrespondences(camera_point,
+                                             target_point,
+                                             target);
+
+  setup.camera_to_target_guess = camera_point;
+
+  PnPResult res = rct_optimizations::optimize(setup);
+
+  EXPECT_TRUE(res.converged);
+  //costs are oddly high for 3d, perfect guess
```

Comment:
This would be an indication that your cost function is not correct. If you put in perfect data, your initial cost per observation should be close to the machine epsilon. You should be seeing error on the order of 1.0e-15. Another way to validate is to make a test where the starting condition is not perfect and see if it converges with low residual error